plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetescluster/KubernetesClusterManagerImpl.java

```diff
@@ -0,0 +1,3061 @@
+            return true;
+        }
+        if (kubernetesCluster.getState().equals(KubernetesCluster.State.Starting)) {
+            LOGGER.debug(String.format("Kubernetes cluster ID: %s is already in starting state", kubernetesCluster.getUuid()));
+            return true;
+        }
+        LOGGER.debug(String.format("Starting Kubernetes cluster ID: %s", kubernetesCluster.getUuid()));
+
+        stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.StartRequested);
+
+        for (final KubernetesClusterVmMapVO vmMapVO : kubernetesClusterVmMapDao.listByClusterId(kubernetesClusterId)) {
+            final UserVmVO vm = userVmDao.findById(vmMapVO.getVmId());
+            try {
+                if (vm == null) {
+                    stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationFailed);
+                    throw new ManagementServerException("Failed to start all VMs in Kubernetes cluster ID: " + kubernetesClusterId);
+                }
+                startKubernetesVM(vm, kubernetesCluster);
+            } catch (ServerApiException ex) {
+                LOGGER.warn("Failed to start VM in Kubernetes cluster ID:" + kubernetesClusterId + " due to " + ex);
+                // dont bail out here. proceed further to stop the reset of the VM's
+            }
+        }
+
+        for (final KubernetesClusterVmMapVO vmMapVO : kubernetesClusterVmMapDao.listByClusterId(kubernetesClusterId)) {
+            final UserVmVO vm = userVmDao.findById(vmMapVO.getVmId());
+            if (vm == null || !vm.getState().equals(VirtualMachine.State.Running)) {
+                stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationFailed);
+                throw new ManagementServerException("Failed to start all VMs in Kubernetes cluster ID: " + kubernetesClusterId);
+            }
+        }
+
+        InetAddress address = null;
+        try {
+            address = InetAddress.getByName(new URL(kubernetesCluster.getEndpoint()).getHost());
+        } catch (MalformedURLException | UnknownHostException ex) {
+            String msg = String.format("Kubernetes cluster ID: %s has invalid API endpoint. Can not verify if cluster is in ready state", kubernetesCluster.getUuid());
+            LOGGER.warn(msg, ex);
+            throw new ManagementServerException(msg, ex);
+        }
+
+        Pair<String, Integer> publicIpSshPort = getKubernetesClusterServerIpSshPort(kubernetesCluster);
+        String publicIpAddress = publicIpSshPort.first();
+        if (Strings.isNullOrEmpty(publicIpAddress)) {
+            String msg = String.format("Failed to start Kubernetes cluster ID: %s as no public IP found for the cluster" , kubernetesCluster.getUuid());
+            LOGGER.warn(msg);
+            stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationFailed);
+            throw new ManagementServerException(msg);
```

Comment:
So not just in the one method. This extra method will make it all the code more readable.